Delay progress bar in Copy-Item and Remove-Item cmdlets

TheSpyGod opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary
I've optimized the Progress Bar of the Copy and Remove commands to display only after the operation lasts more then 200ms, otherwise do the operation without a Progress Bar, I've done this by adding to the IF conditional of IF _copyStopwatch or _removeStopwatch Elapsed time is higher then 0.2 seconds(200ms) add the Progress Bar to the process.
<!-- Summarize your PR between here and the checklist. -->

## PR Context
Fix #23831
<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [ ] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

TheSpyGod committed:
removing progressbar for less then 5 files

I intend to remove the progressbar for less then or equal to 5 file, to make it less anoying TODO: Check for errors when running Remove-Item

TheSpyGod committed:
Changed the progress bar to be 600ms timed

Made it so that the progress bar for removing only appears after the _removeStopwatch reaches 600ms

TheSpyGod committed:
Made both remove and copy display the progress bar after 200ms

Made the copy and remove command display the Progress bar only after the process is lasting more then 200ms

TheSpyGod committed:
Merge pull request #1 from TheSpyGod/issue

Issue

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3111,7 +3112,7 @@ void WriteErrorHelper(Exception exception)
-                        if (_totalFiles > 0)
+                        if (_removeStopwatch.Elapsed.TotalSeconds > 0.2 && _totalFiles > 0)
                         {
                             _removedFiles++;
```
1. This shouldn't be skipped while _removeStopwatch.Elapsed.TotalSeconds <= 0.2
2. I'd make 0.2 C# const since it is used twice.
3. I think 0.2 too small for interactive scenario. I believe user can wait 3-5 seconds before start worrying.

TheSpyGod replied:
Am i understanding correctly, that you want me to make it so that instead of all of it being skipped, the calcuations continue while the instance of the progressBar is created only after 3 - 5 seconds elapsed?

iSazonov replied:
Yes. 
We should do `_removedFiles++` while we don't show progress bar.

TheSpyGod replied:
Got it, i'll send another pull with your suggestions

SteveL-MSFT replied:
Maybe start with 3 seconds?

TheSpyGod committed:
Changed it so that the Progressbar Instance is only created after 3 seconds

Made it so that instead of skipping it all together, only the instance will not be shown until 3 seconds elapse, the calculations will still continue

TheSpyGod committed:
Merge pull request #2 from TheSpyGod/issue

Changed the Creating of the Progress Bar's instance to be after 3 seconds

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -2955,6 +2955,7 @@ protected override void RemoveItem(string path, bool recurse)
 
                 if (Stopping || _removedFiles == _totalFiles)
                 {
+
```
Please remove extra line.
```suggestion
```

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3116,6 +3117,9 @@ void WriteErrorHelper(Exception exception)
                             _removedBytes += fileBytesSize;
                             double speed = _removedBytes / 1024 / 1024 / _removeStopwatch.Elapsed.TotalSeconds;
+
+                            if (_removeStopwatch.Elapsed.TotalSeconds > progressBarDurationThreshold)
```
Please follow our code style guideline - const should start with uppercase letter:
```suggestion
                            if (_removeStopwatch.Elapsed.TotalSeconds > ProgressBarDurationThreshold)
```

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3125,6 +3129,7 @@ void WriteErrorHelper(Exception exception)
                             progress.PercentComplete = percentComplete;
                             progress.RecordType = ProgressRecordType.Processing;
                             WriteProgress(progress);
+                            }
```
Indentation.

SteveL-MSFT replied:
@TheSpyGod it looks like indentation here and below lines 4003-4011 still needs to be addressed

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3116,6 +3117,9 @@ void WriteErrorHelper(Exception exception)
                             _removedFiles++;
                             _removedBytes += fileBytesSize;
                             double speed = _removedBytes / 1024 / 1024 / _removeStopwatch.Elapsed.TotalSeconds;
+
```
The line can be moved in follow `if`.

TheSpyGod committed:
Fixed Indentation as Guidelines Say and Switched Const to PascalCase

Fixed Indentation for IF as Guidelines Say and Switched the Duration const to PascalCase

TheSpyGod committed:
Merge pull request #3 from TheSpyGod/issue

Fixed Indentation, Removed Empty Spaces, and Switched Const to PascalCase

SteveL-MSFT committed:
fix indentation

SteveL-MSFT commented:
@TheSpyGod if you don't want to agree to the CLA, I'll submit a new PR to address this

SteveL-MSFT reviewed (approved):
Fixed the indentation issue

TheSpyGod commented:
@microsoft-github-policy-service agree

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3116,15 +3116,18 @@ void WriteErrorHelper(Exception exception)
-                            progress.PercentComplete = percentComplete;
-                            progress.RecordType = ProgressRecordType.Processing;
-                            WriteProgress(progress);
+                            if (_removeStopwatch.Elapsed.TotalSeconds > ProgressBarDurationThreshold)
```
`double speed = ...` should be moved under follow `if`.

SteveL-MSFT replied:
resolved

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3995,15 +3998,18 @@ private void CopyFileInfoItem(FileInfo file, string destinationPath, bool force,
-                                progress.PercentComplete = percentComplete;
-                                progress.RecordType = ProgressRecordType.Processing;
-                                WriteProgress(progress);
+                                if (_copyStopwatch.Elapsed.TotalSeconds > ProgressBarDurationThreshold)
```
The same.

SteveL-MSFT replied:
resolved

SteveL-MSFT committed:
address Ilya's feedback and reduce threshold to 2 secs

SteveL-MSFT commented:
@iSazonov can you take another look?  I've pushed changes you've requested as I'd like to get this into 7.5 as it annoys me when doing small copies/removes.  I've also reduced the threshold to 2 secs intead of 3.  After building the branch privately, 3 secs seems way too long.

iSazonov reviewed (approved):
@SteveL-MSFT 

> ProgressBarDurationThreshold = 2.0;

Classic definition for interactive scenarios is 5  - 15 seconds. My personal preference is gold middle 10s for the constant. (Something like Exchange Server PowerShell console does.) Otherwise I expect users will see flapped lines. Nevertheless I merge the PR so we get early feedback and adjust the constant before release.

iSazonov commented:
@TheSpyGod Thanks for your contribution.
